File: 工厂方法模式/FactoryLBS/FactoryLBS/ViewController.m
//
//  ViewController.m
//  FactoryLBS
//
//  Created by wanglei on 2019/3/24.
//  Copyright © 2019 wanglei. All rights reserved.
//

#import "ViewController.h"

#import "BaiduMapFactory.h"
#import "GaodeMapFactory.h"
#import "LocationProtocol.h"
#import "MapEngine.h"
#import "MapPointAnnotation.h"

#import <BMKLocationkit/BMKLocationComponent.h>
#import <AMapFoundationKit/AMapFoundationKit.h>

#import <AMapLocationKit/AMapLocationKit.h>

@interface ViewController () <LocationProtocolDelegate, MapViewDelegate>

//@property (nonatomic, strong) BMKLocationManager *locationManager;
//@property (nonatomic, strong) AMapLocationManager *locationManager;
@property (nonatomic, strong) id<LocationProtocol> location;
@property (nonatomic, strong) id<MapViewProtocol> mapView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    
    id<MapFactoryProtocol> factory = [[MapEngine shared] getFactory];
    _mapView = [factory createMapView:self.view.bounds];
    [self.view addSubview:_mapView.mapView];
    _mapView.delegate = self;
    _mapView.showsUserLocation = YES;
    _mapView.userTrackingMode = UserTrackingModeFollowWithHeading;
    
    [_mapView startLocation];
//    _location = [factory getLocationManager];
//    _location.delegate = self;
//    [_location startUpdatingLocation];
//    [_location requestLocationWithReGeocode:YES completionBlock:^(CLLocation * _Nonnull location, id<LocationReGeocode>  _Nonnull regeocode, NSError * _Nonnull error) {
////        NSLog(@"location = %f : %f", location.coordinate.latitude, location.coordinate.longitude);
//
//        MapPointAnnotation *annoation = [[MapPointAnnotation alloc] init];
//        [annoation setCoordinate:location.coordinate];
//        annoation.title = regeocode.district;
//        annoation.subtitle = regeocode.formattedAddress;
//        [mapView addAnnotation:annoation];
//
//        [mapView setCenterLocation:location animated:NO];
//    }];
}

- (void)locationDidUpdateLocation:(CLLocation *)location reGeocode:(id<LocationReGeocode>)reGeocode
{
    [_mapView setCenterLocation:location animated:NO];
}

- (void)locationDidUpdateHeading:(CLHeading *)newHeading
{
    
}

@end

File: 备忘录模式/BankMementoPattern/BankMementoPattern/ViewController.swift
//
//  ViewController.swift
//  BankMementoPattern
//
//  Created by wanglei on 2019/4/9.
//  Copyright © 2019 wanglei. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        let bank = Bank()
        
        bank.addBankEntity(entity: BankEntity(id: 0, name: "aaa", money: 1000))
        bank.addBankEntity(entity: BankEntity(id: 0, name: "bbb", money: 4000))
        bank.addBankEntity(entity: BankEntity(id: 0, name: "ccc", money: 5000))
        let memento = bank.createMemento()
        
        bank.addBankEntity(entity: BankEntity(id: 0, name: "ccc", money: 2000))
        bank.addBankEntity(entity: BankEntity(id: 0, name: "ccc", money: 4000))
        bank.printMoney()
        
        bank.apply(memento: memento) // 备忘录恢复,直接恢复到上次存储点
        bank.printMoney()
    }


}

